File: csharp/triangle/Triangle.cs
using System;
using System.Linq;

public static class Triangle
{
    public static bool IsScalene(double side1, double side2, double side3) =>
        isValidTriangle(new double[]{ side1, side2, side3 })
            && side1 != side2 && side1 != side3 && side2 != side3;

    public static bool IsIsosceles(double side1, double side2, double side3) =>
        isValidTriangle(new double[]{ side1, side2, side3 })
            && (side1 == side2 || side1 == side3 || side2 == side3);

    public static bool IsEquilateral(double side1, double side2, double side3) =>
        side1 != 0 && side1 == side2 && side2 == side3;

    static bool isValidTriangle(double[] sides) {
        var sortedSides = sides.OrderBy(x => x).ToArray();
        return sortedSides[2] > 0 && sortedSides[0] + sortedSides[1] > sortedSides[2];
    }
}

File: c/triangle/src/triangle.c
#include "triangle.h"

bool is_valid_triangle(triangle_t sides)
{
  double a = sides.a;
  double b = sides.b;
  double c = sides.c;

  if (a == 0 || b == 0 || c == 0) {
    return false;
  }

  if (a <= c && b <= c) {
    return a + b > c;
  }
  if (b <= a && c <= a) {
    return b + c > a;
  }
  return a + c > b;
}

bool is_equilateral(triangle_t sides)
{
  return sides.a != 0 && sides.a == sides.b && sides.b == sides.c;
}

bool is_isosceles(triangle_t sides)
{
  return is_valid_triangle(sides) && (
    sides.a == sides.b || sides.a == sides.c || sides.b == sides.c
  );
}
bool is_scalene(triangle_t sides)
{
  return is_valid_triangle(sides) && (
    sides.a != sides.b && sides.a != sides.c && sides.b != sides.c
  );
}
